mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala

```diff
@@ -64,99 +64,105 @@ class Completions(
     case (_: Ident) :: (_: SeqLiteral) :: _ => false
     case _ => true
 
-  /**
-   * It shows the conditions of the code in the cursor position, with
-   * respect to assessing
-   *  1. whether it makes semantic sense to complete
-   *  the position with symbols that are `class`, `trait`,
-   *  `abstract`, `object`, or method
-   *  1. whether it would make syntactic sense to add `[]` `()` or `{}`
-   *  suffixes to the symbols suggested for completion.
-   *
-   *  Note: In the comments `@@` represents the cursor position.
-   *
-   * @param isTypePosition if this is the position for defining a type such as
-   *                       `val a: In@@` or `def i: scala.IndexedSe@@`
-   * @param isNewPosition if it is the position for instantiation after
-   *                      the new keword, as in `new Fil@@`
-   * @param isInstantiationOrMethodCallPos if it is the position for method call or
-   *                                       instantiation as in `object A{ MyCaseClas@@ }`
-   * @param noSquareBracketExists if there is already a square bracket in the position as in
-   *                              `val a: IndexedMa@@[]`
-   */
-  private case class CursorPositionCondition(
-      isTypePosition: Boolean,
-      isNewPosition: Boolean,
-      isInstantiationOrMethodCallPos: Boolean,
-      noSquareBracketExists: Boolean,
-  ):
-
-    def isObjectValidForPos = (!isTypePosition && !isNewPosition)
-
-    /**
-     * classes and traits are type symbols. They are not suitable for
-     *        instantiation or method call positions. Only apply methods and objects
-     *        can be used in such positions.
-     */
-    def isClassOrTraitValidForPos = !isInstantiationOrMethodCallPos
-    def isMethodValidForPos = !isTypePosition
-
-    def isParanethesisValidForPos = !isTypePosition
-    def isSquareBracketsValidForPos = noSquareBracketExists
-      && (isTypePosition || isNewPosition || isInstantiationOrMethodCallPos)
-
-    /**
-     * as in ```new MyTrait@@```
-     */
-    def isCurlyBracesValidForPos = isNewPosition
-  end CursorPositionCondition
-
-  private lazy val cursorPositionCondition =
+  enum CursorPos:
+    case Type(hasTypeParams: Boolean, hasNewKw: Boolean)
+    case Term
+    case Import
```

Comment:
@vzmerr I'm sorry that I missed that thing in your initial PR.
In general this was correct but I reworked it a bit at to operate over a rule that we have 3 types of completions.
It's either a type or term or import.
So, having it we can safely filter out symbols: non type at TypePos, non Term at TermPos. For Import it's ok to reference Term and Type.

Reply:
Thanks for working on this. May I know the cases where we had the regressions?

Reply:
These two:
https://github.com/scalameta/metals/pull/4234/files#diff-9ab9c5f7b86393976b3619ea86ddef73737390ca43e2a3d8ce27d66aa1f576e7R1534-R1570

Reply:
My current PR on object completion also brings substantial change, based on the existing code. Now, if we use this reworked code, how can we work that out?